utils-zookeeper/src/main/java/cz/o2/proxima/utils/zookeeper/ZKGlobalWatermarkTracker.java

```diff
@@ -0,0 +1,458 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.utils.zookeeper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import cz.o2.proxima.storage.watermark.GlobalWatermarkTracker;
+import cz.o2.proxima.util.ExceptionUtils;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.zookeeper.AddWatchMode;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.KeeperException.Code;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.Watcher.Event.EventType;
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooDefs.Ids;
+import org.apache.zookeeper.ZooKeeper;
+
+/** A {@link GlobalWatermarkTracker} that stores global information in Apache Zookeeper. */
+@Slf4j
+public class ZKGlobalWatermarkTracker implements GlobalWatermarkTracker {
+
+  private static final long serialVersionUID = 1L;
+  private static final Instant MAX_WATERMARK = Instant.ofEpochMilli(Long.MAX_VALUE);
+
+  static final String CFG_NAME = "name";
+  static final String ZK_URI = "zk.url";
+  static final String ZK_SESSION_TIMEOUT = "zk.timeout";
+
+  private String trackerName;
+  private String zkConnectString;
+  private String parentNode;
+  private int sessionTimeout;
+  private transient volatile ZooKeeper client;
+
+  @GuardedBy("this")
+  private final Map<String, Long> partialWatermarks = new HashMap<>();
+
+  @VisibleForTesting
+  transient Map<String, Set<CompletableFuture<Void>>> updatedFutures = new ConcurrentHashMap<>();
+
+  private final AtomicLong globalWatermark = new AtomicLong(Long.MIN_VALUE);
+  private transient boolean parentCreated = false;
+  private transient volatile int updateVersion = 0;
+
+  @Override
+  public String getName() {
+    return trackerName;
+  }
+
+  @Override
+  public void configure(Map<String, Object> cfg) {
+    URI uri = getZkUri(cfg);
+    zkConnectString = String.format("%s:%d", uri.getHost(), uri.getPort());
+    sessionTimeout = getSessionTimeout(cfg);
+    trackerName = getTrackerName(cfg);
+    parentNode = normalize(Strings.isNullOrEmpty(uri.getPath()) ? "/" : uri.getPath());
+  }
+
+  @VisibleForTesting
+  static String normalize(String path) {
```

Comment:
can we add javadoc on what kind of normalization we perform?

Reply:
Added comment.